server/google_resources.py

```diff
@@ -31,57 +26,98 @@ def credentials_to_dict(credentials):
+            access_type='offline',
+            include_granted_scopes='true')
+
+        flask.session['state'] = state
     
-    # Create flow instance to manage the OAuth 2.0 Authorization Grant Flow steps.
-    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(CLIENT_SECRETS_FILE, scopes=SCOPES)
-
-    # The URI created here must exactly match one of the authorized redirect URIs
-    # for the OAuth 2.0 client, which you configured in the API Console. If this
-    # value doesn't match an authorized URI, you will get a 'redirect_uri_mismatch'
-    # error.
-
-    flow.redirect_uri = (flask.request.json['redirect_url'] + '/redirect')
-    authorization_url, state = flow.authorization_url(
-        # Enable offline access so that you can refresh an access token without
-        # re-prompting the user for permission. Recommended for web server apps.
-        access_type='offline',
-        # Enable incremental authorization. Recommended as a best practice.
-        include_granted_scopes='true'
-        )
-    # Store the state so the callback can verify the auth server response.
-    flask.session['state'] = state
-    return {"url": authorization_url}, 200
+        return {
+            'url': authorization_url,
+            'state': state
+        }, 200
 
 class OAuth2Callback(Resource):
-  def post(self):
-    # Specify the state when creating the flow in the callback so that it can
-    # verified in the authorization server response.
-    state = flask.session['state']
-    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
-        CLIENT_SECRETS_FILE, scopes=None, state=state)
-    flow.redirect_uri = flask.request.json['redirect_url']
-
-    # Use the authorization server's response to fetch the OAuth 2.0 tokens.
-    authorization_response = flask.request.json["url"]
-    print(flask.request.json)
-    flow.fetch_token(authorization_response=authorization_response)
-
-    # Store credentials in the session.
-    # ACTION ITEM: In a production app, you likely want to save these
-    #              credentials in a persistent database instead.
-    credentials = flow.credentials
-    creds_dict = credentials_to_dict(credentials)
-    save_credentials(creds_dict)
-    return creds_dict, 200
\ No newline at end of file
+    def post(self):
+        state = flask.session['state']
+        
+        flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+            CLIENT_SECRET_FILE, scopes=SCOPES, state=state)
+        flow.redirect_uri = REDIRECT_URI
+        
+        # authorization url
+        authorization_response = flask.request.json['url']
+        flow.fetch_token(authorization_response=authorization_response)
+        
+        credentials = flow.credentials
+        flask.session['credentials'] = credentials_to_dict(credentials)
+        save_credentials(json.dumps(flask.session['credentials'], separators=(',',':')), flask.request.json['userId'])
```

Comment:
store `json.dumps(flask.session['credentials'], separators=(',',':'))` in a variable because it's already used twice here and below

Reply:
Changed to store into a variable on latest commit